### PR TITLE
Make s3d direct-depth range configurable from the dataset (fixes close-range AP3D=0)

### PR DESCRIPTION
**Pull request by Bovey0809**
## Summary

Make the S3D direct-depth range configurable from `depth_min` and `depth_max` in the dataset YAML.

Ownership stays in the existing `DepthDFL.bin_values` buffer: the trainer retargets that buffer once when it builds the model, the loss reads the same buffer for target normalization, decoding already uses it, and checkpoints already persist registered buffers.

Deleted: the trainer callback, wrapper/head resolver, duplicated val/predict range plumbing, generated-reference churn, and unrelated formatting from the earlier branch. The final PR is 5 files (+36/−3) instead of 47 files (+906/−108); parent-branch formatting and generated references were moved to the parent branch that owns them.

The remaining additions are required because the previous 2–80 m range was a module constant with no dataset-owned configuration point; the implementation extends the existing `DepthDFL` buffer instead of adding a second state owner.

## Reproduction

On the unpatched `001-stereo-centernet-gaps` base, a dataset minimum of 0.2 m still decodes the lowest one-hot depth bin at 2.0 m, reproducing the close-range clamp that zeroes the direct-depth signal.

## Validation

- The same lowest-bin case decodes at the configured 0.2 m after range configuration.
- The regression test verifies the loss reads the same 0.2–2.5 m buffer range.
- `pytest tests/test_s3d.py -k 'depth or dfl' --no-cov -q` — 3 passed before the final focused assertion; the final focused test also passes.
- `git diff <base> --check`

## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

### 🌟 Summary
🎯 Adds configurable minimum and maximum depth ranges for S3D models, aligning depth prediction, training loss, datasets, and validation tests.

### 📊 Key Changes
- ➕ Added `depth_min` and `depth_max` options to the KITTI Stereo dataset configuration.
- 🧠 Added `DepthDFL._set_range()` to dynamically update depth bins with validation for valid ranges.
- ⚙️ Updated S3D training to read depth limits from the dataset YAML, with existing defaults as fallback.
- 🔄 Ensured the model’s depth decoder is configured after model creation or weight loading.
- 📉 Updated depth loss normalization to derive its range from the model’s configured depth bins.
- ✅ Added tests verifying custom depth ranges are applied consistently to decoding and loss calculations.

### 🎯 Purpose & Impact
- 🚗 Enables S3D models to support datasets with different usable depth ranges instead of relying on fixed global limits.
- 🎯 Keeps depth prediction and loss calculations synchronized, improving training correctness.
- 🛡️ Prevents invalid configurations by requiring a positive minimum depth below the maximum.
- 🔧 Existing users remain compatible through fallback defaults, while KITTI Stereo now explicitly supports depths from 2 to 80 meters.

### CODE REVIEW

**Comment by UltralyticsAssistant** [comment]
👋 Hello @Bovey0809, thank you for submitting a `ultralytics/ultralytics` 🚀 PR! This is an automated welcome message, and an engineer will assist with the review. To ensure a seamless integration of your work, please review the following checklist:

- ✅ **Define a Purpose**: Clearly explain the purpose of your fix or feature in your PR description, and link to any [relevant issues](https://github.com/ultralytics/ultralytics/issues). Ensure your commit messages are clear, concise, and adhere to the project's conventions.
- ✅ **Synchronize with Source**: Confirm your PR is synchronized with the `ultralytics/ultralytics` `main` branch. If it's behind, update it by clicking the 'Update branch' button or by running `git pull` and `git merge main` locally.
- ✅ **Ensure CI Checks Pass**: Verify all Ultralytics [Continuous Integration (CI)](https://docs.ultralytics.com/help/CI) checks are passing. If any checks fail, please address the issues.
- ✅ **Update Documentation**: Update the relevant [documentation](https://docs.ultralytics.com/) for any new or modified features.
- ✅ **Add Tests**: If applicable, include or update tests to cover your changes, and confirm that all tests are passing.
- ✅ **Sign the CLA**: Please ensure you have signed our [Contributor License Agreement](https://docs.ultralytics.com/help/CLA) if this is your first Ultralytics PR by writing "I have read the CLA Document and I sign the CLA" in a new message.
- ✅ **Minimize Changes**: Limit your changes to the **minimum** necessary for your bug fix or feature addition. _"It is not daily increase but daily decrease, hack away the unessential. The closer to the source, the less wastage there is."_ — Bruce Lee

For more guidance, please refer to our [Contributing Guide](https://docs.ultralytics.com/help/contributing). Don't hesitate to leave a comment if you have any questions. Thank you for contributing to Ultralytics! 🚀

File: ultralytics/models/yolo/s3d/val.py [hunk 1]
```diff
@@ -323,6 +327,17 @@ def _parse_dims(raw_dims):
         if self.std_dims is None:
             LOGGER.info("No std_dims in dataset config, will use defaults")
 
+        # Depth range: dataset YAML may retarget the direct-depth head for close-range scenes.
+        # Apply it to the head so its DepthDFL bins and the decode's variance term share one range,
+        # then read back the effective range (a loaded checkpoint already carries its trained bins).
+        head = unwrap_model(model).model[-1]
```
Comment (UltralyticsAssistant) [inline ultralytics/models/yolo/s3d/val.py]: ⚠️ **HIGH**: This crashes standalone validation (`yolo val`): in that path `model` is an `AutoBackend`, whose `.model` is the `Stereo3DDetModel`, not the layer `ModuleList`. Therefore `unwrap_model(model).model[-1]` attempts to subscript the model object. Unwrap the backend's underlying model before indexing its `.model` layers, while retaining the current direct-model path used during training validation.

File: ultralytics/models/yolo/s3d/val.py [hunk 2]
```diff
@@ -323,6 +327,17 @@ def _parse_dims(raw_dims):
...
+        d_min = self.data.get("depth_min") if hasattr(self, "data") and self.data else None
```
Comment (UltralyticsAssistant) [inline ultralytics/models/yolo/s3d/val.py]: 💡 **MEDIUM**: Standalone validation never receives these values from the dataset YAML: `Stereo3DDetValidator.get_dataset()` currently returns `mean_dims` and `std_dims` but omits `depth_min` and `depth_max`. Consequently, after fixing the AutoBackend access above, `yolo val` will leave the checkpoint/default depth bins in place rather than applying a close-range dataset's configured range. Include both fields in that dataset metadata return.

**Review by UltralyticsAssistant — COMMENTED** [review]
## 🔍 PR Review

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

The training-path range propagation is wired through the head, loss, predictor, and decode variance logic, but standalone S3D validation currently crashes when locating the head and does not retain the dataset depth-range settings.

💬 Posted 2 inline comments

**Review by glenn-jocher — APPROVED** [review]
LGTM at eff4cfc545a2c7ecdbd5f624d1c2811d75c4de19. The unpatched 2.0 m floor was reproduced against a 0.2 m dataset range. The final fix configures the existing `DepthDFL` buffer once, derives loss normalization from that same buffer, and deletes all callback/wrapper/val/predict duplication. Parent formatting is no longer carried by this five-file child PR.

**Comment by UltralyticsAssistant** [comment]
🎉 Thank you and congratulations on merging this impactful PR, @Bovey0809, with contributions from @glenn-jocher!

> “Great things are done by a series of small things brought together.” — Vincent van Gogh

By making S3D depth ranges configurable and keeping decoding, training loss, datasets, and validation synchronized, this work brings greater flexibility and correctness to depth estimation across datasets. The robust validation and backward-compatible defaults make the improvement especially valuable for the Ultralytics community. 🚀
